flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/writer/PulsarWriter.java

```diff
@@ -0,0 +1,203 @@
+        this.pendingMessages = new Semaphore(sinkConfiguration.getMaxPendingMessages());
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void write(IN element, Context context) throws IOException, InterruptedException {
+        // Serialize the incoming element.
+        sinkContextAdapter.updateTimestamp(context);
+        RawMessage<byte[]> message = serializationSchema.serialize(element, sinkContextAdapter);
+
+        // Choose the right topic to send.
+        List<String> availableTopics = metadataListener.availableTopics();
+        String topic = topicRouter.route(element, message, availableTopics, sinkContextAdapter);
+
+        // Create message builder for sending message.
+        TypedMessageBuilder<?> builder = createMessageBuilder(topic, deliveryGuarantee);
+        if (sinkConfiguration.isEnableSchemaEvolution()) {
+            ((TypedMessageBuilder<IN>) builder).value(element);
+        } else {
+            ((TypedMessageBuilder<byte[]>) builder).value(message.getValue());
+        }
+        message.supplement(builder);
+
+        // Perform message sending.
+        if (deliveryGuarantee == DeliveryGuarantee.NONE) {
+            // We would just ignore the sending exception. This may cause data loss.
+            builder.sendAsync();
+        } else {
+            // Waiting for permits to write message.
+            pendingMessages.acquire();
+            CompletableFuture<MessageId> sender = builder.sendAsync();
+            sender.whenComplete(
+                    (id, ex) -> {
+                        pendingMessages.release();
+                        if (ex != null) {
+                            mailboxExecutor.execute(
+                                    () -> {
+                                        throw new FlinkRuntimeException(
+                                                "Failed to send data to Pulsar " + topic, ex);
+                                    },
+                                    "Failed to send message to Pulsar");
+                        } else {
+                            LOG.debug("Sent message to Pulsar {} with message id {}", topic, id);
+                        }
+                    });
+        }
+    }
+
+    private TypedMessageBuilder<?> createMessageBuilder(
+            String topic, DeliveryGuarantee deliveryGuarantee) {
+        Producer<?> producer = producerRegister.getOrCreateProducer(topic);
+        if (deliveryGuarantee == DeliveryGuarantee.EXACTLY_ONCE) {
+            Transaction transaction = producerRegister.getOrCreateTransaction(topic);
+            return producer.newMessage(transaction);
+        } else {
+            return producer.newMessage();
+        }
+    }
+
+    @Override
+    public void flush(boolean endOfInput) throws IOException {
+        while (pendingMessages.availablePermits() < sinkConfiguration.getMaxPendingMessages()) {
```

Comment:
I think you need to flush here if `endOfInput = true` otherwise you might lose records when the job finishes.

Reply:
OK, added.